doc/decisions/operation_sequences.md

```diff
@@ -0,0 +1,143 @@
+# Operation sequences
+
+## Problem
+
+The current public API allows various kind of sequences of `kdbGet` and `kdbSet` calls.
+There is no real guideline of allowed and/or prohibited sequences, other than `kdbGet` for a certain parent key must be called before `kdbSet`:
+
+- initially after `kdbOpen`
+- after conflict errors in `kdbSet`
+
+While researching ways to do change tracking, we discovered that some sequences are currently legal but might be problematic.
+This issue not only effects change tracking, but may cause problems for other use cases too.
+
+Currently, plugins may (and in the case of notification plugins actually do) assume that a `kdbSet` operation for a certain parent key always directly follows the `kdbGet` operation for that same key. 
+However, this is not enforced in any way within Elektra, nor is it documented that it should.
+
+So in reality, as illustrated by the example, `kdbGet` and `kdbSet` of different parent keys might be interwoven.
+
+This is problematic, as plugins may store the result of `kdbGet` and use the stored data in the following `kdbSet` to calculate a changeset.
+Plugins have to do this, as there isn't any other Elektra-provided mechanism currently to do change-tracking for them.
+
+This problem is illustrated by the following example:
+
+```c
+Key * keyA = ...;
+Key * keyB = ...;
+// assume that neither A below B nor B below A
+KeySet * a = ...;
+KeySet * b = ...;
+kdbGet(kdb, a, keyA); // (A)
+kdbGet(kdb, b, keyB); // (B)
+kdbSet(kdb, a, keyA); // (C)
+```
+
+- At point `(A)` the plugins internally store the data below the parent `keyA`.
+- At point `(B)` the same thing happens for the data below the parent `keyB`.
+  This **replaces** the data from `(A)`.
+- At point `(C)` plugins that do change-tracking need to generate a changeset.
+  The changeset will be calculated between the stored data and the new data in the KeySet `a`.
+  Because the stored data is now from `(B)` and **not** from `(A)`, this changeset will be wrong.
+  The calculation will wrongly assume everything in KeySet `b` was removed and everything in KeySet `a` is newly added.
+
+
+To further illustrate the problem, assume KDB contains the following keys:
+
+```
+user:/a/brightness
+user:/a/saturation
+user:/b/background
+user:/b/foreground
+```
+
+`keyA` will be `user:/a`, `keyB` will be `user:/b`.
+Right after `(B)`, the two keysets are filled with the following values:
+
+| `a`                  | `b`                  |
+|----------------------|----------------------|
+| `user:/a/brightness` | `user:/b/background` |
+| `user:/a/saturation` | `user:/b/foreground` |
+
+For a plugin that does change-tracking as described above, the plugin will then wrongly calculate the following changeset at `(C)`:
```

Comment:
> change-tracking as described above

There is not really a change tracking described above.

Please always use same spelling for "change tracking".

Reply:
Yes this was meant.